Support for Matter 1.4 Water Heater device type

lboue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adding support for Matter 1.4 Water Heater :

**Clusters**

| ID (Decimal) | ID (hex) | Name                                                     |
|--------------|----------|----------------------------------------------------------|
|144           |0x0090    |Electrical Power Measurement                              |
|145           |0x0091    |Electrical Energy Measurement                             |
|148           |0x0094    |Water Heater Management                                   |
|152           |0x0098    |Device Energy Management                                  |
|156           |0x009C    |Power Topology                                                     |
|158           |0x009E    |Water Heater Mode                                               |
|159           |0x009F    |Device Energy Management Mode                       |
|513           |0x0201    |Thermostat                                                            |


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- Link to documentation pull request: - https://github.com/home-assistant/home-assistant.io/pull/38514

**Testing**
- [Matter EFR32 Water Heater Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/energy-management-app/silabs)
- [Matter Linux Water Heater Example](https://github.com/project-chip/connectedhomeip/tree/master/examples/energy-management-app/linux)
- [Water Heater App: Interaction using the chip-tool and TestEventTriggers](https://github.com/project-chip/connectedhomeip/tree/master/examples/energy-management-app/linux#water-heater-app-interaction-using-the-chip-tool-and-testeventtriggers)

**WaterHeaterManagement Cluster on Endpoint 2**

**WaterHeaterMode Cluster on Endpoint 2**
 - 3 modes: `[{"0":"Off","1":0,"2":[{"1":16384}]},{"0":"Manual","1":1,"2":[{"1":16385}]},{"0":"Timed","1":2,"2":[{"1":16386}]}]`

![image](https://github.com/user-attachments/assets/8e6b0f27-51ba-4968-a4e1-238a96b6713c)

![image](https://github.com/user-attachments/assets/98d9b329-4482-4d56-adaf-22f47f0a9901)

**Water Heater Mode Cluster**

24-27350-005_Matter-1.4-Application-Cluster-Specification.pdf, page 690 in 9.6. Water Heater Mode Cluster section:
```
9.6.7.1.1. Off Tag
While in modes with this tag, the device will not attempt to keep the water warm.

9.6.7.1.2. Manual Tag
While in modes with this tag, the device will attempt to keep the water warm based on the OccupiedHeatingSetpoint attribute of the associated Thermostat cluster.

9.6.7.1.3. Timed Tag
While in modes with this tag, the device will attempt to keep the water warm based on the Schedules attribute of the associated Thermostat cluster.
```

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

lboue committed:
Create water_heater.json

lboue committed:
Update water_heater.json

lboue committed:
Update water_heater.json

lboue committed:
TankVolume

lboue committed:
TankPercentage

lboue committed:
WaterHeaterMode

WaterHeaterMode

lboue committed:
Update sensor.py

lboue committed:
ruff-format

lboue committed:
Update water_heater.json

 Attributes of WaterHeaterManagement Cluster on Endpoint 2
ClusterId 148 (0x0094)

lboue committed:
Update test_sensor.py

water_heater fixture

lboue committed:
Update test_sensor.py

lboue committed:
SensorDeviceClass=VOLUME_STORAGE for `TankVolume`

lboue committed:
`BoostStateEnum` map

lboue committed:
WaterHeaterManagementBoostState

lboue committed:
Update sensor.py

lboue committed:
WaterHeaterManagementEstimatedHeatRequired

lboue committed:
Fix UnitOfEnergy

lboue committed:
Format

joostlek reviewed (commented):
Why don't we implement the water_heater platform?

lboue committed:
Add `device_types.WaterHeater` to Climate

joostlek commented:
That would be sensors, but I think the main control like modes and stuff should be a `water_heater` rather than a climate and select

lboue commented:
> That would be sensors, but I think the main control like modes and stuff should be a `water_heater` rather than a climate and select

OK but this is a special case, because even if you use `Platform.WATER_HEATER`, you'll have to read the Thermostat cluster, which is usually used by `Platform.CLIMATE`.

lboue committed:
Strings for Tank sensors

lboue committed:
WaterHeater icons

lboue committed:
Update icons.json

lboue committed:
Update strings.json

lboue committed:
Update water_heater.json

lboue committed:
ruff-format

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
Fix tests

lboue committed:
Fix sensor.py

lboue committed:
Fix icons

lboue committed:
WaterHeaterManagementEstimatedHeatRequired

lboue committed:
WaterHeaterManagementBoostState

lboue committed:
BoostState as a binary sensor

lboue committed:
ElectricalPowerMeasurement values

lboue committed:
Fix tests

lboue commented:
I'm trying to integrate the `Platform.WATER_HEATER` platform but so far it's not working: [MatterWaterHeater_testing](https://github.com/lboue/ha_core/blob/MatterWaterHeater_testing/homeassistant/components/matter/water_heater.py).

BDVGitHub commented:
Does this also solve the following problem `Matter smart plug with energy monitoring gets only on/off control in Home Assistant` https://github.com/home-assistant/core/issues/130761 or is this something else?

marcelveldt commented:
> I'm trying to integrate the `Platform.WATER_HEATER` platform but so far it's not working: [MatterWaterHeater_testing](https://github.com/lboue/ha_core/blob/MatterWaterHeater_testing/homeassistant/components/matter/water_heater.py). The entity is displayed but remains in the unknown state.

Please adjust the PR here so we can give feedback that way or help you getting it over the finish line like we did before.

lboue committed:
Create water_heater.py

lboue committed:
Update climate.py from dev branch

lboue committed:
Resolve conflicts

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
ruff-format

lboue committed:
Add Platform.WATER_HEATER

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
Update water_heater.py

lboue committed:
Update water_heater.py

lboue committed:
Update water_heater.py

lboue committed:
Update water_heater.py

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
Add WaterHeaterManagement sensors

lboue committed:
Update tests

lboue committed:
Add select test

lboue committed:
Add strings

lboue committed:
First try with water_heater

lboue committed:
Testing current_operation

lboue committed:
BoostState attribute

lboue committed:
target_temperature attributes

lboue committed:
target_temperature attribute

lboue committed:
set_temperature and set_operation_mode

lboue committed:
turn_on / turn_off

lboue committed:
Trigger Boost command

lboue committed:
Fix WaterHeaterBoostInfoStruct

lboue committed:
Add test file

lboue committed:
Add climate cluster to fixture

lboue committed:
Add climate cluster to fixture

lboue committed:
Add tests

lboue committed:
Add ON_OFF feature

lboue committed:
Update tests

lboue commented:
I've now used the `water_heater` entity and associated it with the EP2 `Thermostat` cluster.
Request an initial review to assess the improvements to be made.

lboue committed:
Update tests

lboue commented:
What would be the best way to select the length of the boost period? We'd need to be able to read this value when boost mode is activated.

NoRi2909 commented on homeassistant/components/matter/strings.json:
```diff
@@ -267,6 +270,15 @@
+        "name": "Tank percentage"
+      },
+      "estimated_heat_required": {
+        "name": "Heat energy needed"
```
Correct me if I'm wrong:
If that is the estimated energy needed to heat the water that should be "Required heating energy"

"heat energy" sounds more like the energy contained in hot water.

lboue replied:
I agree with you. I will fix it.

lboue committed:
Translate WaterHeaterMode

lboue committed:
Change description

lboue committed:
Update test and snapshots

lboue committed:
Update snapshots

joostlek commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,205 @@
+        )
+        system_mode_value = WATER_HEATER_SYSTEM_MODE_MAP.get(operation_mode)
+        if system_mode_value is None:
+            raise ValueError(f"Unsupported hvac mode {operation_mode} in Matter")
```
We should raise `ServiceValidationError` here

joostlek commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,205 @@
+        self._attr_target_temperature = self._get_temperature_in_degrees(
+            clusters.Thermostat.Attributes.OccupiedHeatingSetpoint
+        )
+        BoostState = self.get_matter_attribute_value(
```
snake_case

joostlek commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,205 @@
+        BoostState = self.get_matter_attribute_value(
+            clusters.WaterHeaterManagement.Attributes.BoostState
+        )
+        if (BoostState) == clusters.WaterHeaterManagement.Enums.BoostStateEnum.kActive:
```
not sure why the parenthesis

joostlek commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,205 @@
+
+# Discovery schema(s) to map Matter Attributes to HA entities
+DISCOVERY_SCHEMAS = [
+    MatterDiscoverySchema(
+        platform=Platform.WATER_HEATER,
+        entity_description=WaterHeaterEntityDescription(
+            key="MatterWaterHeater",
+        ),
+        entity_class=MatterWaterHeater,
+        required_attributes=(
+            clusters.Thermostat.Attributes.OccupiedHeatingSetpoint,
+            clusters.Thermostat.Attributes.AbsMinHeatSetpointLimit,
+            clusters.Thermostat.Attributes.AbsMaxHeatSetpointLimit,
+            clusters.Thermostat.Attributes.LocalTemperature,
+            clusters.WaterHeaterManagement.Attributes.FeatureMap,
+        ),
+        optional_attributes=(
+            clusters.WaterHeaterManagement.Attributes.HeaterTypes,
+            clusters.WaterHeaterManagement.Attributes.BoostState,
+            clusters.WaterHeaterManagement.Attributes.HeatDemand,
+        ),
+        device_type=(device_types.WaterHeater,),
+        allow_multi=True,  # also used for sensor entity
+    ),
```
We probably have to set name to `None` to make the device name be the name of the entity, as I am assuming the water heater always is the main entity of the device

lboue committed:
Set entity name to None to make the device name be the name of the entity

lboue committed:
Format

lboue committed:
Update water_heater.py

lboue committed:
Fix format

lboue committed:
ruff-format

lboue committed:
Import ServiceValidationError

lboue committed:
Merge branch 'dev' into WaterHeater

joostlek commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -106,8 +107,9 @@ async def async_set_operation_mode(self, operation_mode: str) -> None:
-        await self.write_attribute(
+            raise ServiceValidationError(
+                f"Unsupported hvac mode {operation_mode} in Matter"
+            )        await self.write_attribute(
```
```suggestion
            )
        await self.write_attribute(
```

lboue committed:
Update homeassistant/components/matter/water_heater.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

lboue committed:
Update water_heater.py

lboue committed:
Update test_water_heater.py

lboue committed:
Update test_water_heater.ambr

lboue committed:
Update test_water_heater.py

lboue committed:
Update select.py

lboue committed:
Update snapshots

lboue committed:
Rename to boost_info

lboue committed:
Set WaterHeaterMode

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
Update snapshots

lboue committed:
Update snapshots

lboue committed:
Fix for warning
W7431: Argument 3 should be of type AddConfigEntryEntitiesCallback in async_setup_entry (hass-argument-type)

lboue commented:
I have just corrected the warnings and errors.

NoRi2909 commented on homeassistant/components/matter/strings.json:
```diff
@@ -300,6 +315,11 @@
+    },
+    "water_heater": {
+      "water_heater": {
+        "name": "Water heater"
```
For consistency with the "name" reference for "valve" you could replace this with

`[%key:component::water_heater::title%]`

lboue committed:
Update strings.json

lboue commented on homeassistant/components/matter/strings.json:
```diff
@@ -270,6 +276,15 @@
+        "name": "Tank volume"
+      },
+      "tank_percentage": {
+        "name": "Tank percentage"
```
Yes. From specs (Matter-1.4-Application-Cluster-Specification.pdf):
**TankPercentage Attribute**
This attribute SHALL indicate an approximate level of hot water stored in the tank, which might help consumers understand the amount of hot water remaining in the tank. The accuracy of this attribute is manufacturer specific.
In most hot water tanks, there is a stratification effect where the hot water layer rests above a cooler layer of water below. For this reason cold water is fed in at the bottom of the tank and the hot water is drawn from the top.
Some water tanks might use multiple temperature probes to estimate the level of the hot water layer. A water heater with multiple temperature probes is likely to implement an algorithm to esti
mate the hot water tank percentage by taking into account the temperature values of each probe to determine the height of the hot water.

NoRi2909 replied:
Thanks for the detailed explanation!

Yes, I think "Hot water level" makes most sense then and is easy to translate.

"Percentage" is very common in English but as the level is shown in percent I vote for dropping it from the entity names and use the space for more detailed names.

lboue committed:
Update strings and tests

lboue committed:
Merge branch 'dev' into WaterHeater

lboue committed:
Fix missing brace

lboue committed:
Update tests

lboue committed:
fix test

lboue committed:
Updates strings

lboue committed:
Fix async_set_temperature

lboue committed:
Update tests

lboue committed:
Update tests

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+        """Set new target temperature."""
+        target_temperature: float | None = kwargs.get(ATTR_TEMPERATURE)
+        if target_temperature is not None:
+            if self.target_temperature != target_temperature:
```
Combine the two checks on line 87 and 88 with a logical operator eg and.

If we invert the conditions and return if true, we can outdent below.

lboue replied:
Fixed

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+                    value=int(target_temperature * TEMPERATURE_SCALING_FACTOR),
+                    matter_attribute=matter_attribute,
+                )
+            return
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+        ] = clusters.WaterHeaterManagement.Structs.WaterHeaterBoostInfoStruct(
+            duration=3600
+        )
+        system_mode_value = WATER_HEATER_SYSTEM_MODE_MAP.get(operation_mode)
```
```suggestion
        system_mode_value = WATER_HEATER_SYSTEM_MODE_MAP[operation_mode]
```

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+            duration=3600
+        )
+        system_mode_value = WATER_HEATER_SYSTEM_MODE_MAP.get(operation_mode)
+        if system_mode_value is None:
```
We don't need this check. It's already checked in the parent water heater entity class.

lboue replied:
Removed

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        # self._calculate_features()
```
Commented code.

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+            ),
+        )
+        self._attr_target_temperature_low = self._attr_min_temp
+        self._attr_target_temperature_high = self._attr_max_temp
```
The max and min temperatures do not mean the same thing as max and min target temperatures.

lboue replied:
Removed

MartinHjelmare commented on homeassistant/components/matter/select.py:
```diff
@@ -318,6 +319,18 @@ def _update_from_device(self) -> None:
+        ),
+        entity_class=MatterModeSelectEntity,
+        required_attributes=(
+            clusters.WaterHeaterMode.Attributes.CurrentMode,
```
What mode does this entity change?

MartinHjelmare replied:
I think we need to think about how to explain the different entities that change mode to users. We probably at least need to extend the documentation regarding this.

lboue replied:
Like this? https://deploy-preview-38514--home-assistant-docs.netlify.app/integrations/matter/#matter-appliances-modes

MartinHjelmare replied:
We should list the modes yes, but we also need to explain the other modes on the water heater entity and the relationship.

But let's first decide if this is the way to go forward or not. I'll get back to you about this.

lboue replied:
WHMode selector removed as requested.

lboue committed:
Update homeassistant/components/matter/water_heater.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

lboue committed:
Sort strings in strings.json

lboue committed:
Update homeassistant/components/matter/water_heater.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

lboue committed:
Remove unused line

lboue committed:
Remove min/max target temperatures

lboue committed:
Remove BOOST_STATE_MAP

lboue commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+
+WATER_HEATER_SYSTEM_MODE_MAP = {
+    STATE_ECO: 4,
+    STATE_HIGH_DEMAND: 4,
```
The aim is to apply the correct value to the ThermostatRunningMode attribute of the Thermostat cluster. Given that in the Matter standard, the water heater is considered as a Thermostat.
So both eco and high demand (boost) correspond to the heat mode of the Thermostat cluster.

lboue commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,206 @@
+    STATE_OFF: 0,
+}
+
+BOOST_STATE_MAP = {
```
It is used in async_set_operation_mode()

lboue committed:
Add comment

lboue committed:
Remove SUPPORT_FLAGS_HEATER

lboue committed:
Remove system_mode_value check

lboue committed:
Update homeassistant/components/matter/water_heater.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

lboue committed:
Reformat async_set_temperature()

lboue committed:
Merge branch 'dev' into WaterHeater

lboue commented:
I searched but couldn't find why I have an error as pytest works locally:

```
pytest tests/components/matter/test_water_heater.py --snapshot-update --log-level=DEBUG
Test session starts (platform: linux, Python 3.13.1, pytest 8.3.5, pytest-sugar 1.0.0)
rootdir: /home/ludovic/HomeAssistant/ha_core
configfile: pyproject.toml
plugins: syrupy-4.8.1, aiohttp-1.1.0, respx-0.22.0, anyio-4.9.0, asyncio-0.26.0, cov-6.0.0, requests-mock-1.12.1, xdist-3.6.1, socket-0.7.0, pytest_freezer-0.4.9, github-actions-annotate-failures-0.3.0, unordered-0.6.1, picked-0.5.1, sugar-1.0.0, timeout-2.3.1
asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 48 items                                                                                                                                                                                                                  

 tests/components/matter/test_water_heater.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                        100% ██████████
------------------------------------------------------------------------------------------------------ snapshot report summary ------------------------------------------------------------------------------------------------------
2 snapshots passed.

Results (17.72s):
      48 passed

```

lboue committed:
Update snapshots

marcelveldt reviewed (approved):
Excellent work @lboue !

lboue committed:
Remove MatterWaterHeaterMode selector

lboue committed:
Update snapshots

lboue commented:
> We'll discuss this in the next core meeting, how we should present and explain the different mode selections to users.

WHMode selector removed as requested.

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+    await hass.services.async_call(
+        "water_heater",
+        "set_operation_mode",
+        {"entity_id": "water_heater.water_heater", "operation_mode": STATE_OFF},
```
I suggest we parametrize the operation_mode value and the corresponding matter attribute value and test all the possible modes.

MartinHjelmare replied:
```suggestion
        {"entity_id": "water_heater.water_heater", "operation_mode": operation_mode},
```

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+
+
+@pytest.mark.parametrize("node_fixture", ["silabs_water_heater"])
+async def test_water_heater_service_calls(
```
```suggestion
async def test_water_heater_set_temperature(
```

lboue replied:
Fixed

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+    matter_client: MagicMock,
+    matter_node: MatterNode,
+) -> None:
+    """Test water_heater platform service calls."""
```
```suggestion
    """Test water_heater set temperature service."""
```

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+    matter_node: MatterNode,
+) -> None:
+    """Test water_heater platform service calls."""
+    # test single-setpoint temperature adjustment when boost mode is active
```
This comment doesn't seem to match the test. Boost mode isn't active when the eco operation mode is active.

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+    )
+    matter_client.write_attribute.reset_mock()
+
+    # test change mode to off
```
```suggestion

@pytest.mark.parametrize("node_fixture", ["silabs_water_heater"])
@pytest.mark.parametrize(
    ("operation_mode", "matter_attribute_value"),
    [
        (STATE_OFF, 0)
        (STATE_ECO, 4)
        (STATE_HIGH_DEMAND, 4)
    ],
)
async def test_water_heater_set_operation_mode(
    hass: HomeAssistant,
    matter_client: MagicMock,
    matter_node: MatterNode,
    operation_mode: str,
    matter_attribute_value: int,
) -> None:
    """Test water_heater set operation mode service."""
    state = hass.states.get("water_heater.water_heater")

    assert state
    assert state.state == STATE_ECO

```

lboue replied:
Fixed

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+            endpoint_id=2,
+            attribute=clusters.Thermostat.Attributes.SystemMode,
+        ),
+        value=0,
```
```suggestion
        value=matter_attribute_value,
```

lboue replied:
Fixed

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+        ),
+        value=0,
+    )
+    matter_client.send_device_command.reset_mock()
```
```suggestion
```

lboue replied:
Fixed

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,115 @@
+    assert state.attributes["supported_features"] & mask == mask
+
+    # test BoostState update from device
+    set_node_attribute(matter_node, 2, 148, 5, 1)
```
We should test all the mode updates and that they generate the correct entity state.

MartinHjelmare replied:
We could do that after calling the set_operation_mode service in the last test, where I've made a suggestion how to parametrize the operation mode and attribute value.

lboue replied:
Move after calling the set_operation_mode service in the last test

lboue committed:
Rename test to test_water_heater_set_temperature

lboue committed:
Add test_water_heater_set_operation_mode

lboue committed:
Remove reset_mock

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -95,21 +89,45 @@ async def test_water_heater_set_temperature(
+        )
+
+        # test BoostState update from device
+        set_node_attribute(matter_node, 2, 148, 5, 1)
```
We want to test the corresponding attribute as set by the operation mode service call.

lboue replied:
The aim here is to test an attribute change not from the HA side but on the Matter device side. For example, if boost mode is activated by pressing a button on the device interface.

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,189 @@
+        self._update_from_device()
+        # Trigger Boost command
+        if operation_mode == STATE_HIGH_DEMAND:
+            await self.send_device_command(
```
We need to assert this command in a test sent when selecting the high demand mode.

lboue replied:
Fixed

MartinHjelmare commented on homeassistant/components/matter/water_heater.py:
```diff
@@ -0,0 +1,189 @@
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on water heater."""
+        await self.async_set_operation_mode("eco")
```
On/off service calls aren't tested. Please add tests for that.

lboue replied:
Fixed

MartinHjelmare reviewed (commented):
We seem to be missing tests for some features of the water heater. Make sure to test everything.

lboue committed:
Update tests/components/matter/test_water_heater.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

lboue committed:
Add test_update_from_water_heater

lboue commented:
> We seem to be missing tests for some features of the water heater. Make sure to test everything.

I try to do my best. I'm not very familiar with python unit tests.

lboue committed:
Add test_water_heater_turn_on_off

lboue committed:
Add test_water_heater_boostmode

lboue committed:
Merge branch 'dev' into WaterHeater

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -129,12 +129,73 @@ async def test_water_heater_set_operation_mode(
+        )
+
+        state = hass.states.get("water_heater.water_heater")
+        assert state.state == STATE_HIGH_DEMAND
```
I don't believe it matters here what the state is, since we haven't triggered any state update with a subscription callback. The service action call itself doesn't change the state, right?

Check all the tests for this pattern if we should update it.

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,244 @@
+                endpoint_id=2,
+                attribute=clusters.Thermostat.Attributes.SystemMode,
+            ),
+            value=0,
```
How can this be 0 here when it's 4 in the `test_water_heater_set_operation_mode` test?

MartinHjelmare commented on tests/components/matter/test_water_heater.py:
```diff
@@ -0,0 +1,244 @@
+        set_node_attribute(matter_node, 2, 148, 5, 1)
+        await trigger_subscription_callback(hass, matter_client)
+        state = hass.states.get("water_heater.water_heater")
+        assert state.state == STATE_HIGH_DEMAND
```
We need to test all possible states updates, from the different operation modes.

lboue committed:
Fix SystemMode value for STATE_HIGH_DEMAND

lboue committed:
Add disable boost from water heater device side test

lboue committed:
Remove unused lines

lboue committed:
Remove unused lines

MartinHjelmare committed:
Fix test indentation

MartinHjelmare committed:
Fix water heater tests

MartinHjelmare committed:
Check for None

lboue commented:
> I fixed the water heater tests and a check for the temperature where we want to allow zero but not None before scaling.

Thanks

MartinHjelmare reviewed (approved):
Thanks!
